### ISSUE TITLE
@import CSS Rule Not Top-Level

### ISSUE THREAD

**Issue by TimFeldmann-Patriot** [marco-prontera/vite-plugin-css-injected-by-js#70]
When the style tag is created, any CSS `@import` rules have the opportunity to not be created at the top of the style tag. This causes the import rule to not fetch the associated CSS and therefore not be included in the final style sheet.

This can be reproduced using the following code:

```
var A = document.createElement("style");
A.textContent = 
`
div {
    background-color: blue;
}

@import "https://YourCssFileHere.css";
`;

document.head.appendChild(A);
```

Is there anyway we can force all @import rules to be specified at the top of the style tag?

Thanks.

**Comment by TimFeldmann-Patriot** [comment]
I was able to workaround this issue by writing a custom `injectCodeFunction` to move `@import` rules to the beginning of the style tag:

```
        cssInjectedByJsPlugin({
          injectCodeFunction: function fixImport(cssCode, options) {
              try {
                  if (typeof document != 'undefined') {
                      const importRegex = /@import.*?(?<=;)/g;
                      const importRules = cssCode.match(importRegex);

                      cssCode = cssCode.replace(importRegex, "");
                      cssCode = `${importRules?.join("")}${cssCode}`;

                      const elementStyle = document.createElement('style');
                      elementStyle.appendChild(document.createTextNode(`${cssCode}`));

                      document.head.appendChild(elementStyle);
                  }
              } catch (e) {
                  console.error('vite-plugin-css-injected-by-js', e);
              }
          }
      })
```

**Comment by marco-prontera** [comment]
Hi @TimFeldmann-Patriot can I know how many CSS files you have in output without this plugin?
I think this problem should be only present when many CSS files will be exported. Maybe you can solve that problem without writing this function but by changing some other vite's configuration.

**Comment by TimFeldmann-Patriot** [comment]
@marco-prontera The `@import` rule is actually coming from an NPM package which grabs styles from a CDN, I do not have direct source access to the npm package's CSS file which is declaring the `@import`.

I was hoping there would be an easy way to re-order the CSS during Vite build so that this regex would not need to be performed by each browser visiting the SPA. I have used Vite's manual chunks to manually split JS chunks for my single-spa repository in the past, and I saw how you were extracting the final CSS [here](https://github.com/marco-prontera/vite-plugin-css-injected-by-js/blob/main/src/index.ts#L46).  I figured this Regex could be performed during build around line 57.

I thought about forking this repo as a short-term solution as well. Since this plugin creates a single style tag, the problem will always be that any CSS modules which declare an `@import` will end up somewhere below the top of the style tag, no matter how it is declared within each individual CSS module.

**Comment by marco-prontera** [comment]
@TimFeldmann-Patriot Published the [v2.4.0](https://www.npmjs.com/package/vite-plugin-css-injected-by-js?activeTab=readme) in the new version you can find documentation about the `preRenderCSSCode` parameter that enables plugin users to customize the CSS code during the build phase of the injection code.

I need more time to think and to see which is the best way to integrate the logic you proposed. It's right I know but manipulating CSS code is a hard thing to do without breaking something. If you have time to help me with a PR I really appreciate it, in the meanwhile thank you for your contribution!

**Comment by TimFeldmann-Patriot** [comment]
@marco-prontera Thank you! The `preRenderCSSCode` function is perfect!

I understand the reservations on including this regex directly in the plugin. Although I tested my regex with CSS spanning thousands of lines, I wonder if there is some condition which would match text that should not be matched. Perhaps if a class name includes '@import` this regex would break.

I will think of a fool-proof way to solve this and if I find something I'll make a PR.

Thank you again!
